cmd/podman/compose.go

```diff
@@ -0,0 +1,264 @@
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+var composeCommand = &cobra.Command{
+	Use:   "compose [options]",
+	Short: "Run compose workloads via an external provider such as docker-compose or podman-compose",
+	Long: `This command is a thin wrapper around an external compose provider such as docker-compose or podman-compose.  This means that podman compose is executing another tool that implements the compose functionality but sets up the environment in a way to let the compose provider communicate transparently with the local Podman socket.  The specified options as well the command and argument are passed directly to the compose provider.
+
+The default compose providers are docker-compose and podman-compose.  If installed, docker-compose takes precedence since it is the original implementation of the Compose specification and is widely used on the supported platforms (i.e., Linux, Mac OS, Windows).
+
+If you want to change the default behavior or have a custom installation path for your provider of choice, please change the compose_provider field in containers.conf(5).  You may also set PODMAN_COMPOSE_PROVIDER environment variable.`,
+	RunE:              composeMain,
+	ValidArgsFunction: completion.AutocompleteNone,
+	Example: `podman compose -f nginx.yaml up --detach
+  podman --log-level=debug compose -f many-images.yaml pull`,
+	DisableFlagParsing: true,
+	Annotations:        map[string]string{registry.ParentNSRequired: ""}, // don't join user NS for SSH to work correctly
+}
+
+func init() {
+	// NOTE: we need to fully disable flag parsing and manually parse the
+	// flags in composeMain. cobra's FParseErrWhitelist will strip off
+	// unknown flags _before_ the first argument.  So `--unknown argument`
+	// will show as `argument`.
+
+	registry.Commands = append(registry.Commands, registry.CliCommand{Command: composeCommand})
+}
+
+// composeProvider provides the name of or absolute path to the compose
+// provider (i.e., the external binary such as docker-compose).
+func composeProvider() (string, error) {
+	if value, ok := os.LookupEnv("PODMAN_COMPOSE_PROVIDER"); ok {
+		return value, nil
+	}
+
+	candidates := registry.PodmanConfig().ContainersConfDefaultsRO.Engine.ComposeProviders
+	if len(candidates) == 0 {
+		return "", errors.New("no compose provider specified, please refer to `man podman-compose` for details")
+	}
+
+	for _, candidate := range candidates {
+		path, err := exec.LookPath(candidate)
+		if err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				continue
+			}
+			return "", err
+		}
+		// First specified provider "candidate" wins.
+		logrus.Debugf("Found compose provider %q", path)
+		return path, nil
+	}
+
+	return "", errors.New("no configured compose provider found on system, please refer to `man podman-compose` for details")
+}
+
+// composeEnv returns the compose-specific environment variables.
+func composeEnv() ([]string, error) {
+	hostValue, err := func() (string, error) {
```

Comment:
Why is this a anonymous function? Can you split this out to a proper function, I find it hard to follow the logic with such long anonymous functions.